### ISSUE TITLE
Strictly decode to case class disallowing extraneous fields in JSON

### ISSUE THREAD

**Issue by LeifW** [argonaut-io/argonaut#104]
So that 

``` scala
case class Person(name: String)
Parse.decode[Person]("""{"name": "Bob", "foo": "dunno"}""")
```

 will yield something about "Error: unexpected field "foo"".

Is this feasible?

Example use case: When people call our API, it would be helpful to tell them if the misspelled a key or added junk that doesn't do anything - tell them that's not key that means anything to us and it'll be ignored / rejected as invalid.

**Comment by lukasz-golebiewski** [comment]
@markhibberd The URL you pasted is dead. Do you have any examples on how to use `validate`?

**Comment by kevinmeredith** [comment]
To make sure that I understand this problem, @LeifW, let's consider an example.

Given:

```
import argonaut._, Argonaut._

case class Person(name: String)

implicit def decode: DecodeJson[Person] =
  DecodeJson ( c => 
    for {
      name <- (c --\ "name").as[String]
    } yield Person(name)
  )

scala> Parse.decode[Person]("""{"name": "Bob", "foo": "dunno"}""")
res5: Either[Either[String,(String, argonaut.CursorHistory)],Person] = Right(Person(Bob))
```

you would expect, either by default in `Parse#decode` or an optional `Parse#decodeExactly`, that the extra `foo` field should result in a `Left` rather than `Right`?

**Comment by LeifW** [comment]
Exactly, yes, thanks for the example. I guess I'd lean towards the optional
decodeExactly; I think people are used to the present behavior. A use case
I was thinking of was better error messages. For large complex structures
it can be easy to misspell a field, and the "missing field foo" would be
easier to figure out when accompanied by a "unexpected field foob". Also
sometimes people pass extra fields in their JSON API calls, and then they
just get cargo culted in successive generations, even though they never did
anything - people are afraid to remove them for fear of breaking something.
I'd like to head that kind of thing off explicitly at the beginning.

**Comment by seanparsons** [comment]
`DecodeJson.validate` should be able to achieve that (albeit at a cost to the caller). It does chime slightly with an idea I've had in the past for a different way to handle parsing completely but behaviour like this would sit weirdly in our current model for decoding.

**Comment by kevinmeredith** [comment]
Thanks, @seanparsons, for pointing me to that method!

Locally, I added the following:

`case Some(w) => println(a); DecodeResult.ok(w)`

to this [line](https://github.com/argonaut-io/argonaut/blob/master/argonaut/shared/src/main/scala/argonaut/DecodeJson.scala#L171).

Then, I re-ran the above example:

## JSON contains exact # of fields, one, as case class to which it's decoded

```
scala> Parse.decode[Person]("""{"name": "Bob"}""")
HCursor(CObject(CJson({"name":"Bob"}),false,object[(name,"Bob")],(name,"Bob")),CursorHistory(List(El(CursorOpDownField(name),true))))
res0: Either[Either[String,(String, argonaut.CursorHistory)],Person] = Right(Person(Bob))
```

## JSON has one extra field, i.e. not in the Person

```
scala> Parse.decode[Person]("""{"name": "Bob", "foo": "dunno"}""")
HCursor(CObject(CJson({"name":"Bob","foo":"dunno"}),false,object[(name,"Bob"),(foo,"dunno")],(name,"Bob")),CursorHistory(List(El(CursorOpDownField(name),true))))
res1: Either[Either[String,(String, argonaut.CursorHistory)],Person] = Right(Person(Bob))
```

After decoding, i.e. `JSON => Person`, is it possible to figure out if the `*Cursor` has any extra, i.e. not walked, fields?

I looked at the scaladocs for `HCursor` and `ACursor`, but I'm not sure. Please help me out?

**Comment by seanparsons** [comment]
@kevinmeredith Not walked? I _think_ you'd be better off just checking that `fieldSet` matches what you expect.

**Comment by kevinmeredith** [comment]
> matches what you expect

Could you please say more?

Taking your reply, @seanparsons, I posted this StackOverflow question - http://stackoverflow.com/questions/43433054/retrieving-traversed-json-fields. 

I appreciate any help, please.

**Comment by seanparsons** [comment]
This is what you want to check against correct?

```
scala> json.right.get.hcursor.fieldSet
res1: Option[Set[argonaut.Json.JsonField]] = Some(Set(foo))
```

**Comment by kevinmeredith** [comment]
@seanparsons Given:

```scala
case class Person(name: String)
```

and:

```
scala> Parse.parse("""{"name": "Bob", "foo": "dunno"}""").toEither.right.get.hcursor.fieldSet
res6: Option[Set[argonaut.Json.JsonField]] = Some(Set(name, foo))
```

How would you use `fieldSet` to determine if there's any extraneous/extra fields in the input JSON?

Also, my original thought was:

1.  Get list of fields not walked/traversed - http://stackoverflow.com/questions/43433054/retrieving-traversed-json-fields
2. Decode JSON to case class
3. if step 2 succeeds, check if 1 is empty. If it's non-empty, then return an error that there's extra fields

Thanks for your continued help.

**Comment by seanparsons** [comment]
Well surely the `fieldSet` you want is `Some(Set("name"))` right?

Unless I'm mistaken that means in the `DecodeJson.validate` call the function `HCursor => Boolean` would look something like `_.fieldSet == Some(Set("name"))`.

To support a `decodeExactly` we'd need to do a non-trivial amount of changes because of the way the existing code is written.

**Comment by kevinmeredith** [comment]
Given all of the fields of a `case class A` and all of the `Json`'s keys, assuming we decode the `Json` to an `A`, we can compare `A`'s fields to `Json`'s keys to determine this `decodeExactly` behavior, no?

https://stackoverflow.com/a/27445097/409976 shows how, using `shapeless`, we can get all of a case classes' fields w/ `LabelledGeneric`. 

With respect to the second problem, i.e. getting all keys from a `Json`, I'm not sure if such a method exists. I don't believe that `Cursor#fields` would work, as is:

```
scala> Json.obj(
     "a" -> jString(""), 
     "b" -> jBool(false), 
     "c" -> Json.obj("a" -> jNull) 
).cursor.fields
res27: Option[List[argonaut.Json.JsonField]] = Some(List(a, b, c))
```

Does a `allFields` method, which would find all fields from the `Json` recursively, exist? If not, perhaps a PR could implement that method, and then, using `LabelledGeneric` and the aforementioned method, we could address this issue?

Finally, if you all find this approach to be technically sound, then perhaps it belongs in https://github.com/alexarchambault/argonaut-shapeless?

**Comment by seanparsons** [comment]
@kevinmeredith The `fields` and `fieldsSet` methods gives you the fields of the object currently in focus. So if you want to check that it has only fields a, b and c you can validate against that.

Hence this:

```scala
scala> val testDecoder: DecodeJson[(String, Boolean)] = jdecode2L((a: String, b: Boolean) => (a, b))("a", "b")
testDecoder: argonaut.DecodeJson[(String, Boolean)] = argonaut.DecodeJson$$anon$4@3f59fa4e

scala> val improvedTestDecoder = testDecoder.validate(_.fieldSet == Some(Set("a", "b")), "Wrong fields")
improvedTestDecoder: argonaut.DecodeJson[(String, Boolean)] = argonaut.DecodeJson$$anon$4@45ed447c

scala> improvedTestDecoder.decodeJson(Json("a" := "q", "b" := true))
res7: argonaut.DecodeResult[(String, Boolean)] = DecodeResult(Right((q,true)))

scala> improvedTestDecoder.decodeJson(Json("a" := "q", "b" := true, "c" := 1))
res8: argonaut.DecodeResult[(String, Boolean)] = DecodeResult(Left((Wrong fields,CursorHistory(List()))))
```

**Comment by kevinmeredith** [comment]
As I understand your reply, @seanparsons, you believe the `LabelledGeneric` approach that I proposed in https://github.com/argonaut-io/argonaut/issues/104#issuecomment-313884079 to be unnecessary given the solution that you've provided with `DecodeJson#validate`?

If so, perhaps I should ask if the application of `LabelledGeneric` would be appropriate to https://github.com/alexarchambault/argonaut-shapeless, i.e. to derive such a `DecodeJson` without the need to call `DecodeJson#validate` manually?

**Comment by seanparsons** [comment]
@kevinmeredith I was pointing out the manual solution, a generic solution would likely want to invoke the same kind of code. It would definitely be a nice thing to have in `argonaut-shapeless` but that's not really a question for me! :)

**Comment by alexarchambault** [comment]
That's definitely do-able with argonaut-shapeless. It has some `JsonProductCodec`, that can be customized to check for any non-decoded field. One should just roughly adapt back [these lines](https://github.com/alexarchambault/plotly-scala/blob/85c53585cda618e96aca94f03c7e46b318feecb6/render/shared/src/main/scala/plotly/Codecs.scala#L167-L209) to argonaut-shapeless, and ensure the implicit they define is in scope when the `DecodeJson` are derived. Then decoding would fail if any extra field is present.
